custom_components/localtuya/light.py

```diff
@@ -120,16 +204,28 @@ async def async_turn_on(self, **kwargs):
                 self._lower_brightness,
                 self._upper_brightness,
             )
-            await self._device.set_dp(brightness, self._config.get(CONF_BRIGHTNESS))
+            if self.light_mode == LIGHT_MODES[1]:
+                hs_color = self.hs_color
+                hs_color = (hs_color[0], hs_color[1], int(kwargs[ATTR_BRIGHTNESS]))
+                hs_color = self.hsv_to_tuya_hsv(hs_color)
+                await self._device.set_dp(hs_color, self._config.get(CONF_HS_COLOR))
+            else:
+                await self._device.set_dp(brightness, self._config.get(CONF_BRIGHTNESS))
 
-        if ATTR_HS_COLOR in kwargs:
-            raise ValueError(" TODO implement RGB from HS")
+        if ATTR_HS_COLOR in kwargs and (features & SUPPORT_COLOR):
+            hs_color = (kwargs[ATTR_HS_COLOR][0], kwargs[ATTR_HS_COLOR][1], self.brightness)
+            hs_color = self.hsv_to_tuya_hsv(hs_color)
+            if self.light_mode != LIGHT_MODES[1]:
+                await self._device.set_dp(LIGHT_MODES[1], self._config.get(CONF_LIGHT_MODE))
+            await self._device.set_dp(hs_color, self._config.get(CONF_HS_COLOR))
 
         if ATTR_COLOR_TEMP in kwargs and (features & SUPPORT_COLOR_TEMP):
-            color_temp = int(
-                255
-                - (255 / (MAX_MIRED - MIN_MIRED))
-                * (int(kwargs[ATTR_COLOR_TEMP]) - MIN_MIRED)
+            color_temp = map_range_inverted(
+                int(kwargs[ATTR_COLOR_TEMP]),
+                MIN_MIRED,
+                MAX_MIRED,
+                self._lower_color_temp,
+                self._upper_color_temp,
             )
             await self._device.set_dp(color_temp, self._config.get(CONF_COLOR_TEMP))
```

Comment:
```suggestion
            if self.light_mode != LIGHT_MODES[0]:  #If not white mode
                await self._device.set_dp(LIGHT_MODES[0], self._config.get(CONF_LIGHT_MODE))
            await self._device.set_dp(color_temp, self._config.get(CONF_COLOR_TEMP))
```
The light mode must be set to "white" when setting color temperature. Otherwise the light will remain in color mode despite the fact that color temperature has been changed (which means we want white light).

Reply:
I originally set the light to white but I noticed that for my model it was not necessary. Anyhow sounds good 👍

Reply:
The differences between different models make everything a bit complex, that's true. For my model for instance, it is necessary to change the mode.